keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,445 @@
+-->
+
+## Motivation
+
+Some workloads need all pods to be running at the same time to make progress, for
+example, when they require pod-to-pod communication. In that case a large
+workload may have difficulty scheduling all of its pods for a prolonged time as
+smaller workloads would continue to block the slots. Moreover, a pair of
+workloads may deadlock if the amount of actual resources available is below the
+quota configured to determine admission.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- a mechanism to ensure that a job gets assigned physical resources when
+unsuspended by Kueue
+- a timeout on reaching the `PodsReady` condition by a workload since unsuspending
+of the corresponding Job
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- guarantee that two workloads would not schedule pods concurrently. Example
+scenarios in which two workloads may still concurrently schedule their pods:
+  - when succeeded pods are replaced with new because job's parallelism is less than its completions;
+  - when a failed pod gets replaced
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+We introduce a mechanism to ensure workloads get their physical resources
+assigned by avoiding concurrent scheduling of workloads. More precisely, we
+block admission of new workloads until the first batch of the Job is scheduled.
+Additionally, we await with starting the Job until all other admitted workloads
+have their pods scheduled. This behavior can be opted-in by a Job by a new
+annotation.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a machine learning researcher I want to run a workload, comprising tens of
+pods that all need to run at the same time. Autoscaling the cluster and
+cheduling of all pods in this workload take a couple of minutes and other
```

Comment:
I'm not sure about 'scheduling' and 'can occupy the resources my workloads could use otherwise'. Does it assume that some workloads schedule outside of Kueue? If everything goes through Kueue and ClusterQueue is configured consistently with capacity, scheduling should be still correct (the problem is with autoscaling/provisioning, or other start issues beyond ClusterQueue control).

Btw. do we envision other start issues than provisioning? Do we want to handle them? We sort of do with the proposed solution. It may have consequences for the naming. I considered atomic-provisioning=true, but since it covers other start issues, maybe ensure-pods-start is more appropriate and direct.